pontoon/pretranslation/pretranslate.py

```diff
@@ -1,130 +1,257 @@
+        match entity.resource.format:
+            case "ftl":
+                self.format = Format.fluent
+            case "po":
+                self.format = Format.mf2
+            case _:
+                self.format = None
+        self.source = entity.string
+        self.locale = locale
+        self.preserve_placeables = preserve_placeables
+        self.services = []
 
-        return (pretranslation, author)
+    def message(self, msg: Message) -> None:
+        """Modifies `msg`."""
+        if isinstance(msg, PatternMessage):
+            msg.pattern = self.pattern(msg.pattern)
+        else:
+            # catchall category is always last, and is always included
+            plurals = self.locale.cldr_plurals_list()[:-1]
 
-    # For now, disable pretranslation for plural gettext messages.
-    # https://github.com/mozilla/pontoon/issues/3694
-    elif entity.resource.format == "po" and ".match" in source:
-        return None
+            # Plural selectors need special attention
+            plural_selectors = [
+                idx
+                for idx, sel in enumerate(msg.selector_expressions())
+                if sel.function in ("integer", "number")
+            ]
 
-    else:
-        pretranslation, service = get_pretranslated_data(
-            source, locale, preserve_placeables
+            # Do not translate plural "one" variants if not used in the target locale
+            has_pc_one = "one" in plurals
+            rm_keys = []
+            for keys, pattern in msg.variants.items():
+                if not has_pc_one and any(
+                    keys[idx] == "one" for idx in plural_selectors
+                ):
+                    rm_keys.append(keys)
+                else:
+                    pattern[:] = self.pattern(pattern)
+            for keys in rm_keys:
+                del msg.variants[keys]
+
+            # Copy catchall patterns for the locale's other plural categories
+            if plurals:
+                for idx in plural_selectors:
+                    tgt_variants = {}
+                    for keys, pattern in msg.variants.items():
+                        key = keys[idx]
+                        if isinstance(key, CatchallKey):
+                            for pc in plurals:
+                                pc_keys = keys[:idx] + (pc,) + keys[idx + 1 :]
+                                if pc_keys not in msg.variants:
+                                    tgt_variants[pc_keys] = deepcopy(pattern)
+                        tgt_variants[keys] = pattern
+                    msg.variants = tgt_variants
+
+    def pattern(self, pattern: Pattern) -> Pattern:
+        # First try to get a 100% match from Translation Memory
+        tm_source = (
+            "".join(
+                el.value
+                if isinstance(el, FTL.TextElement)
+                else serialize_expression(el)
+                for el in fluent_astify_message(PatternMessage(pattern)).elements
+            )
```

Comment:
This smells like a use case for `moz.l10n.message.serialize_message` to support Fluent pattern serialization, which it doesn't yet.

Reply:
File a boog?